packages/query-core/src/utils.ts

```diff
@@ -353,6 +353,19 @@ export function replaceData<
   if (typeof options.structuralSharing === 'function') {
     return options.structuralSharing(prevData, data) as TData
   } else if (options.structuralSharing !== false) {
+    if (process.env.NODE_ENV !== 'production') {
+      try {
+        JSON.stringify(prevData)
+        JSON.stringify(data)
```

Comment:
This now throws an error when data contains BigInt. But it seems like react-query supports having BigInt in the query result. Should I create an issue for this?

Reply:
I'm think it's still good to log the error because structuralSharing creates an overhead and you should likely turn it off if you have something that isn't serializable. But we should likely remove the throw. Logging is enough. Please file a PR

Reply:
Thank you @Efimenko for raising that PR 🙌 
I have a question. From what I see in [`replaceEqualDeep`](https://github.com/TanStack/query/blob/8af72729f0457289e675816453ba9d2d2ed4f604/packages/query-core/src/utils.ts#L243-L281), there's nothing preventing BigInt comparison to break. For example, `123n === 123n` will be `true`. Is there anything else I'm missing where `BigInt` could break`this.setData()`?

Reply:
interesting. I didn't know that BigInt was referentially equal, but not json serializable. According to [the documentation](https://tanstack.com/query/latest/docs/framework/react/guides/render-optimizations#structural-sharing), we only support json serializable values in `structuralSharing`, so I think the warning is still warranted.

Reply:
Yeah, this seems to have affected a lot of upstream Wagmi users. Wonder if this should have been a breaking change.

Reply:
It's "just" a dev mode warning. Why would it be breaking?

Reply:
Apologies. Didn't see [this commit](https://github.com/TanStack/query/commit/fb9d4f6138645a4812d7590a39476f880d181b40#diff-9701bc1ef4ed4515a08de71beb3a01aac6de69115ceda88ad67995f1d5feab8cL366).

Reply:
@TkDodo – would you be open to a PR that checked for cyclic references instead of checking for non-serializable properties via `JSON.stringify`? Feel like a lot of consumers would be invoking query functions that return `BigInt`s, so it might be a bit odd seeing these warning messages flying around when `replaceEqualDeep` is also compatible with `BigInt`s.

Reply:
> Wonder if it’s worth conveying that structural sharing is moreso only compatible with referentially stable values instead of JSON-serializable values.

I simply wasn't aware that `replaceEqualDeep` works with `BigInt` - our general recommendation is to use json serializable values. If we say BigInt is supported, then why isn't Map supported, or Set, or Date, or ... it's a slippery slope. Also, in queryKeys, we actually use `JSON.stringify`, so if you use a BigInt there, it will fail.

Why is JavaScript so weird that BigInt is referentially stable across values, but not serializable lol. 

>  would you be open to a PR that checked for cyclic references instead of checking for non-serializable values via JSON.stringify?

yeah I guess that's fine. I think just calling `replaceEqualDeep` itself in a try/catch would just work, too ?

Reply:
done! https://github.com/TanStack/query/pull/8030